pkg/reconciler/common/prune.go

```diff
@@ -38,58 +41,200 @@ const (
 	JobsTKNImageName = "IMAGE_JOB_PRUNER_TKN"
 	ownerAPIVer      = "operator.tekton.dev/v1alpha1"
 	ownerKind        = "TektonConfig"
+	pruneSkip        = "operator.tekton.dev/prune.skip"
+	pruneResources   = "operator.tekton.dev/prune.resources"
+	pruneStrategy    = "operator.tekton.dev/prune.strategy"
+	pruneKeep        = "operator.tekton.dev/prune.keep"
+	pruneKeepSince   = "operator.tekton.dev/prune.keep-since"
+	pruneSchedule    = "operator.tekton.dev/prune.schedule"
+	pruneCronLabel   = "tektonconfig.operator.tekton.dev/pruner"
+	keep             = "keep"
+	keepSince        = "keep-since"
 )
 
-func Prune(k kubernetes.Interface, ctx context.Context, tC *v1alpha1.TektonConfig) error {
+type Pruner struct {
+	kc       kubernetes.Interface
+	tknImage string
+}
+type pruneConfigPerNS struct {
+	config v1alpha1.Prune
+}
+
+// all the namespaces of default and the annotationbased
+type pruningNs struct {
+	commonScheduleNs map[string]*pruneConfigPerNS
+	uniqueScheduleNS map[string]*pruneConfigPerNS
+}
 
-	if len(tC.Spec.Pruner.Resources) == 0 || tC.Spec.Pruner.Schedule == "" {
-		return checkAndDelete(k, ctx, tC.Spec.TargetNamespace)
+func Prune(ctx context.Context, k kubernetes.Interface, tC *v1alpha1.TektonConfig) error {
+	pruner := &Pruner{kc: k}
+	//may be reconciled by removing pruning spec from tektonConfig
+	if pruner.removedFromTektonConfig(tC.Spec.Pruner) {
+		return pruner.checkAndDeleteCron(ctx, tC.Spec.TargetNamespace)
 	}
 
-	tknImage := os.Getenv(JobsTKNImageName)
-	if tknImage == "" {
+	annotationRemovedNamespaces, err := pruner.checkAnnotationsRemovedNamespaces(ctx, tC.Spec.TargetNamespace)
+	if err != nil {
+		return err
+	}
+	if len(annotationRemovedNamespaces) > 0 {
+		for _, ns := range annotationRemovedNamespaces {
+			if err := pruner.checkAndDeleteCron(ctx, ns); err != nil {
+				return err
+			}
+		}
+	}
+
+	tknImageFromEnv := os.Getenv(JobsTKNImageName)
+	if tknImageFromEnv == "" {
 		return fmt.Errorf("%s environment variable not found", JobsTKNImageName)
 	}
-	pru := tC.Spec.Pruner
+	pruner.tknImage = tknImageFromEnv
 	logger := logging.FromContext(ctx)
 	ownerRef := v1.OwnerReference{
```

Comment:
here we can use 
```
ownerRef := *metav1.NewControllerRef(config, config.GetGroupVersionKind())
```

Reply:
Yes, we discussed to add it in iterative pr but I think we can update it here itself.